### PR TITLE
Move syntax::ext to a syntax_expand and refactor some attribute logic

### PR DESCRIPTION
**Pull request by Centril**
Part of https://github.com/rust-lang/rust/pull/65324.

r? @petrochenkov

### CODE REVIEW

File: src/libsyntax_ext/util.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+use syntax_pos::Symbol;
```
Comment (Centril) [inline src/libsyntax_ext/util.rs]: I know this is a tiny module but it doesn't really fit anywhere else except maybe `lib.rs` but that file is pretty much only about "register all the things" which the function below is not about.

File: src/test/ui-fulldeps/auxiliary/roman-numerals.rs [hunk 2]
```diff
@@ -10,13 +10,14 @@
 #![feature(plugin_registrar, rustc_private)]
```
Comment (Centril) [inline src/test/ui-fulldeps/auxiliary/roman-numerals.rs]: Maybe let's remove these tests? They are pretty annoying...

File: src/test/ui-fulldeps/ast_stmt_expr_attr.rs [hunk 3]
```diff
@@ -18,7 +18,7 @@ use syntax::parse::new_parser_from_source_str;
 use syntax::parse::parser::Parser;
```
Comment (Centril) [inline src/test/ui-fulldeps/ast_stmt_expr_attr.rs]: Shouldn't this be a normal ui test?

File: src/test/ui-fulldeps/auxiliary/roman-numerals.rs (hunk 2 above)
Comment (petrochenkov) [reply src/test/ui-fulldeps/auxiliary/roman-numerals.rs]: These are smoke tests for syntactic plugins, the only thing making sure that they don't break.
I think they should exist until we remove the plugins themselves.

File: src/test/ui-fulldeps/ast_stmt_expr_attr.rs (hunk 3 above)
Comment (petrochenkov) [reply src/test/ui-fulldeps/ast_stmt_expr_attr.rs]: AFAIK, anything depending on compiler crates has to be a fulldeps test.

Comment (Centril) [reply src/test/ui-fulldeps/ast_stmt_expr_attr.rs]: Yeah, what I wonder tho if this test could not be expressed as a series of UI tests instead -- the structure seems to be just ordinary compile-pass / compile-fail wrt. "reject" and "check". Using `cfg(FALSE)` it seems to me we can migrate this test to use that instead and privatize some stuff in the process.

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -522,7 +520,7 @@ impl<'a> Parser<'a> {
 
     /// If the next token is the given keyword, eats it and returns `true`.
     /// Otherwise, returns `false`. An expectation is also added for diagnostics purposes.
-    pub fn eat_keyword(&mut self, kw: Symbol) -> bool {
+    fn eat_keyword(&mut self, kw: Symbol) -> bool {
```
Comment (topecongiro) [inline src/libsyntax/parse/parser.rs]: @Centril Please reconsider before making these kinds of 'refactorings.' Tools like rustfmt which rely on the libsyntax suffer severely from reduced visibility. Now I have to iterate through the changes, re-publish hidden fields and methods, make a PR and wait until it gets merged.

I understand that the libsyntax is primarily for the rustc itself, and its interface is unstable. But please be conservative about the visibility of its interface.

Comment (Centril) [reply src/libsyntax/parse/parser.rs]: @topecongiro If you rely on something then there should be a comment to that effect in the rustc sources. Reducing visibility is an important tool for cleaning up the rustc internals and should be used.

**Pull request merged**